### PR TITLE
ISPN-5414 Support named parameters in DSL based queries

### PR DESCRIPTION
**Pull request by anistor**
jira: https://issues.jboss.org/browse/ISPN-5414
      https://issues.jboss.org/browse/ISPN-5705

### CODE REVIEW

**Comment by anistor** [comment]
This is a first cut. Adds named param functionality. A bit more work will follow to make this useable for query caching.

File: query/src/main/java/org/infinispan/query/continuous/JPAContinuousQueryCacheEventFilterConverter.java [hunk 1]
```diff
@@ -120,14 +126,34 @@ public String toString() {
...
          output.writeUTF(filterAndConverter.jpaQuery);
+         Map<String, Object> namedParameters = filterAndConverter.namedParameters;
+         if (namedParameters != null) {
+            output.writeInt(namedParameters.size());
+            for (Map.Entry<String, Object> e : namedParameters.entrySet()) {
+               output.writeUTF(e.getKey());
+               output.writeObject(e.getValue());
+            }
+         } else {
+            output.writeInt(-1);
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/continuous/JPAContinuousQueryCacheEventFilterConverter.java]: ```UnsignedNumeric.writeUnsignedInt``` and ```0``` when no params ?

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/JPAFilterAndConverter.java [hunk 2]
```diff
@@ -116,14 +126,34 @@ public String toString() {
...
+         Map<String, Object> namedParameters = filterAndConverter.namedParameters;
+         if (namedParameters != null) {
+            output.writeInt(namedParameters.size());
+            for (Map.Entry<String, Object> e : namedParameters.entrySet()) {
+               output.writeUTF(e.getKey());
+               output.writeObject(e.getValue());
+            }
+         } else {
+            output.writeInt(-1);
+         }
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/dsl/embedded/impl/JPAFilterAndConverter.java]: The as commented previously: can you write an unsigned int or  0 when no params present?

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java [hunk 3]
```diff
@@ -417,27 +443,34 @@ public JPAFilterAndConverter run() {
...
+   //todo [anistor] test this method with non-indexed fields and see if a proper exception is thrown
    /**
     * Build a Lucene index query.
     */
-   //todo [anistor] make this private when LuceneQuery/LuceneQueryBuilder are removed
-   public LuceneQuery buildLuceneQuery(QueryFactory queryFactory, String jpqlString, long startOffset, int maxResults) {
+   private BaseQuery buildLuceneQuery(QueryFactory queryFactory, String jpqlString, Map<String, Object> namedParameters, String[] projections, long startOffset, int maxResults) {
+      EmbeddedLuceneQuery query = new EmbeddedLuceneQuery(this, queryFactory, jpqlString, namedParameters, projections, startOffset, maxResults);
+      if (namedParameters != null) {
+         query.setParameters(namedParameters);    //todo [anistor] hack ?
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java]: wdym by "hack" here?

**Comment by gustavonalle** [comment]
looks nice :+1:

File: query/src/main/java/org/infinispan/query/continuous/JPAContinuousQueryCacheEventFilterConverter.java (hunk 1 above)
Comment (anistor) [inline query/src/main/java/org/infinispan/query/continuous/JPAContinuousQueryCacheEventFilterConverter.java]: Done

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/JPAFilterAndConverter.java (hunk 2 above)
Comment (anistor) [inline query/src/main/java/org/infinispan/query/dsl/embedded/impl/JPAFilterAndConverter.java]: Done

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAContinuousQueryProtobufCacheEventFilterConverter.java [hunk 4]
```diff
@@ -145,14 +151,34 @@ private ContinuousQueryResult filterAndConvert(Object key, Object oldValue, Obje
...
+         Map<String, Object> namedParameters = filterAndConverter.namedParameters;
+         if (namedParameters != null) {
+            output.writeInt(namedParameters.size());
+            for (Map.Entry<String, Object> e : namedParameters.entrySet()) {
+               output.writeUTF(e.getKey());
+               output.writeObject(e.getValue());
+            }
+         } else {
+            output.writeInt(-1);
+         }
```
Comment (gustavonalle) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAContinuousQueryProtobufCacheEventFilterConverter.java]: same as previous comments

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAProtobufFilterAndConverter.java [hunk 5]
```diff
@@ -59,11 +61,32 @@ public String toString() {
...
          output.writeUTF(filterAndConverter.getJPAQuery());
+         Map<String, Object> namedParameters = filterAndConverter.getNamedParameters();
+         if (namedParameters != null) {
+            output.writeInt(namedParameters.size());
+            for (Map.Entry<String, Object> e : namedParameters.entrySet()) {
+               output.writeUTF(e.getKey());
+               output.writeObject(e.getValue());
+            }
+         } else {
+            output.writeInt(-1);
```
Comment (gustavonalle) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAProtobufFilterAndConverter.java]: same as previous comment

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java (hunk 3 above)
Comment (anistor) [inline query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java]: was a leftover. removed it.

**Comment by anistor** [comment]
Updated.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAContinuousQueryProtobufCacheEventFilterConverter.java (hunk 4 above)
Comment (anistor) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAContinuousQueryProtobufCacheEventFilterConverter.java]: Done

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAProtobufFilterAndConverter.java (hunk 5 above)
Comment (anistor) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/filter/JPAProtobufFilterAndConverter.java]: Done

**Comment by gustavonalle** [comment]
pulling

**Comment by anistor** [comment]
Updated again :)

**Comment by gustavonalle** [comment]
re-pulling

**Pull request merged**